IPython/extensions/autoreload.py

```diff
@@ -427,6 +469,57 @@ def __call__(self):
 ]
 
 
+class ImportFromTracker:
+    def __init__(self, imports_froms, symbol_map):
+        self.imports_froms = imports_froms
+        # symbol_map maps original_name -> list of resolved_names
+        self.symbol_map = {}
+        if symbol_map:
+            for module_name, mappings in symbol_map.items():
+                self.symbol_map[module_name] = {}
+                for original_name, resolved_names in mappings.items():
+                    if isinstance(resolved_names, list):
+                        self.symbol_map[module_name][original_name] = resolved_names[:]
+                    else:
+                        self.symbol_map[module_name][original_name] = [resolved_names]
+        else:
+            self.symbol_map = symbol_map or {}
+
+    def add_import(self, module_name, original_name, resolved_name):
```

Comment:
New code should be typed so that the typechecker runs on it. My guess would be:

```suggestion
    def add_import(self, module_name: str, original_name: str, resolved_name: str) -> None:
```